Sources/Commands/SwiftScriptTool.swift

```diff
@@ -0,0 +1,374 @@
+            try localFileSystem.removeFileTree(cacheDirPath)
+        }
+    }
+
+    struct Update: ScriptCommand {
+        static let configuration = CommandConfiguration(
+            abstract: "Update package dependencies")
+
+        @OptionGroup(_hiddenFromHelp: true)
+        var swiftOptions: SwiftToolOptions
+
+        @OptionGroup()
+        var options: ScriptToolOptions
+
+        @Flag(name: [.long, .customShort("n")],
+              help: "Display the list of dependencies that can be updated")
+        var dryRun: Bool = false
+
+        @Argument(help: "The packages to update")
+        var packages: [String] = []
+
+        func run(_ swiftTool: SwiftTool, as productName: String, at cacheDirPath: AbsolutePath) throws {
+            let workspace = try swiftTool.getActiveWorkspace()
+
+            let changes = try workspace.updateDependencies(
+                root: swiftTool.getWorkspaceRoot(),
+                packages: packages,
+                diagnostics: swiftTool.diagnostics,
+                dryRun: dryRun
+            )
+
+            // try to load the graph which will emit any errors
+            if !swiftTool.diagnostics.hasErrors {
+                _ = try workspace.loadPackageGraph(
+                    rootInput: swiftTool.getWorkspaceRoot(),
+                    diagnostics: swiftTool.diagnostics
+                )
+            }
+
+            if let pinsStore = swiftTool.diagnostics.wrap({ try workspace.pinsStore.load() }),
+               let changes = changes, dryRun {
+                logPackageChanges(changes: changes, pins: pinsStore)
+            }
+
+            if !dryRun {
+                // Throw if there were errors when loading the graph.
+                // The actual errors will be printed before exiting.
+                guard !swiftTool.diagnostics.hasErrors else {
+                    throw ExitCode.failure
+                }
+            }
+        }
+    }
+}
+
+extension SwiftScriptTool {
+    struct List: ParsableCommand {
+        static let configuration = CommandConfiguration(
+            abstract: "List script caches")
+
+        @OptionGroup(_hiddenFromHelp: true)
+        var swiftOptions: SwiftToolOptions
+
+        func run() throws {
+            let cacheDir = try localFileSystem.getOrCreateSwiftScriptCacheDirectory()
+            let scripts = try localFileSystem.getDirectoryContents(cacheDir)
+            let resolved = try scripts.map { script -> (String, AbsolutePath) in
+                let sourceDir = cacheDir.appending(components: script, "Sources")
+                guard localFileSystem.isDirectory(sourceDir),
+                      let name = try localFileSystem.getDirectoryContents(sourceDir).first,
+                      case let realpath = resolveSymlinks(sourceDir.appending(components: name, "main.swift")) else {
+                          throw Diagnostics.fatalError
```

Comment:
Should this really end the enumeration, or should the entry be either skipped or listed as a stale entry?

Reply:
🤔You're right, better skipping instead of failing here.